// bits-ui/dist/bits/context-menu/ctx.js
import { createBitAttrs, getOptionUpdater, removeUndefined } from "../../internal/index.js";
import { createContextMenu } from "@melt-ui/svelte";
import { getContext, setContext } from "svelte";
import { getPositioningUpdater } from "../floating/helpers.js";
const NAME = "menu";
const PARTS = [
    "arrow",
    "checkbox-indicator",
    "checkbox-item",
    "content",
    "group",
    "item",
    "label",
    "radio-group",
    "radio-item",
    "separator",
    "sub-content",
    "sub-trigger",
    "trigger"
];
export const getAttrs = createBitAttrs("menu", PARTS);
export function setCtx(props) {
    const contextMenu = createContextMenu(removeUndefined(props));
    setContext(NAME, contextMenu);
    return {
        ...contextMenu,
        updateOption: getOptionUpdater(contextMenu.options)
    };
}
export function getCtx() {
    return getContext(NAME);
}
const defaultPlacement = {
    side: "bottom",
    align: "start"
};
export function updatePositioning(props) {
    const withDefaults = { ...defaultPlacement, ...props };
    const { options: { positioning } } = getCtx();
    const updater = getPositioningUpdater(positioning);
    updater(withDefaults);
}
