// bits-ui/dist/bits/link-preview/ctx.js
import { createBitAttrs, getOptionUpdater, removeUndefined } from "../../internal/index.js";
import { createLinkPreview } from "@melt-ui/svelte";
import { getContext, setContext } from "svelte";
import { getPositioningUpdater } from "../floating/helpers.js";
const NAME = "link-preview";
const PARTS = ["arrow", "content", "trigger"];
export const getAttrs = createBitAttrs(NAME, PARTS);
export function getCtx() {
    return getContext(NAME);
}
export function setCtx(props) {
    const linkPreview = createLinkPreview({
        ...removeUndefined(props),
        forceVisible: true
    });
    setContext(NAME, linkPreview);
    return {
        ...linkPreview,
        updateOption: getOptionUpdater(linkPreview.options)
    };
}
export function setArrow(size = 8) {
    const linkPreview = getCtx();
    linkPreview.options.arrowSize.set(size);
    return linkPreview;
}
const defaultPlacement = {
    side: "bottom",
    align: "center"
};
export function updatePositioning(props) {
    const withDefaults = { ...defaultPlacement, ...props };
    const { options: { positioning } } = getCtx();
    const updater = getPositioningUpdater(positioning);
    updater(withDefaults);
}
