// bits-ui/dist/bits/popover/ctx.js
import { createPopover } from "@melt-ui/svelte";
import { getContext, setContext } from "svelte";
import { createBitAttrs, getOptionUpdater, removeUndefined } from "../../internal/index.js";
import { getPositioningUpdater } from "../floating/helpers.js";
const NAME = "popover";
const PARTS = ["arrow", "close", "content", "trigger"];
export const getAttrs = createBitAttrs(NAME, PARTS);
export function setCtx(props) {
    const popover = createPopover({
        ...removeUndefined(props),
        forceVisible: true
    });
    setContext(NAME, popover);
    return {
        ...popover,
        updateOption: getOptionUpdater(popover.options)
    };
}
export function getCtx() {
    return getContext(NAME);
}
export function setArrow(size = 8) {
    const popover = getCtx();
    popover.options.arrowSize.set(size);
    return popover;
}
const defaultPlacement = {
    side: "bottom",
    align: "center"
};
export function updatePositioning(props) {
    const withDefaults = { ...defaultPlacement, ...props };
    const { options: { positioning } } = getCtx();
    const updater = getPositioningUpdater(positioning);
    updater(withDefaults);
}
