// bits-ui/dist/bits/toolbar/ctx.js
import { createBitAttrs, getOptionUpdater, removeUndefined } from "../../internal/index.js";
import { createToolbar } from "@melt-ui/svelte";
import { getContext, setContext } from "svelte";
const NAME = "toolbar";
const GROUP_NAME = "toolbar-group";
const PARTS = ["root", "button", "link", "group", "group-item"];
export const getAttrs = createBitAttrs(NAME, PARTS);
export function setCtx(props) {
    const toolbar = createToolbar(removeUndefined(props));
    setContext(NAME, toolbar);
    return {
        ...toolbar,
        updateOption: getOptionUpdater(toolbar.options)
    };
}
export function setGroupCtx(props) {
    const { builders: { createToolbarGroup } } = getCtx();
    const group = createToolbarGroup(removeUndefined(props));
    setContext(GROUP_NAME, group);
    return {
        ...group,
        updateOption: getOptionUpdater(group.options)
    };
}
export function getCtx() {
    return getContext(NAME);
}
export function getGroupCtx() {
    return getContext(GROUP_NAME);
}
