// @evidence-dev/plugin-connector/dist/loadConfig/load-config.js
var $1CGwI$chalk = require("chalk");
var $1CGwI$fs = require("fs");
var $1CGwI$yaml = require("yaml");
var $1CGwI$zod = require("zod");


function $parcel$interopDefault(a) {
  return a && a.__esModule ? a.default : a;
}

function $parcel$export(e, n, v, s) {
  Object.defineProperty(e, n, {get: v, set: s, enumerable: true, configurable: true});
}

$parcel$export(module.exports, "loadConfig", () => $3956dd9640f2b520$export$c1a4367d4847eb06);
/**
 * Renames the '_errors' property to 'errors' in the given object and its nested objects recursively.
 * It also removes any empty errors arrays
 *
 * @param {any} obj - The object to rename the '_errors' property in.
 * @return {Object} The object with the renamed property.
 */ function $63f219b28f6851a4$export$71e7d3deffa0730b(obj) {
    for(const key in obj){
        if (typeof obj[key] === "object") $63f219b28f6851a4$export$71e7d3deffa0730b(obj[key]); // recursively traverse nested objects
        if (key === "_errors") {
            if (obj["_errors"].length) // De-duplicate
            obj["errors"] = Array.from(new Set(obj["_errors"]));
            delete obj["_errors"];
        }
    }
    return obj;
}






const $7b0083e9d7491326$export$797917169e2ec068 = (0, $1CGwI$zod.z).object({
    overrides: (0, $1CGwI$zod.z).array((0, $1CGwI$zod.z).string()).default([]),
    aliases: (0, $1CGwI$zod.z).record((0, $1CGwI$zod.z).string({
        description: "Component Name"
    }), (0, $1CGwI$zod.z).string({
        description: "Alias to apply"
    })).default({}),
    provides: (0, $1CGwI$zod.z).array((0, $1CGwI$zod.z).string()).default([])
});
const $7b0083e9d7491326$export$81ab33ffc20a47da = (0, $1CGwI$zod.z).object({
    overrides: (0, $1CGwI$zod.z).array((0, $1CGwI$zod.z).string()).default([])
});
const $7b0083e9d7491326$export$e691085fbd9bf5be = (0, $1CGwI$zod.z).object({
    components: (0, $1CGwI$zod.z).record((0, $1CGwI$zod.z).string(), $7b0083e9d7491326$export$797917169e2ec068),
    datasources: (0, $1CGwI$zod.z).record((0, $1CGwI$zod.z).string({
        description: "Plugin Package Name"
    }), $7b0083e9d7491326$export$81ab33ffc20a47da).default({})
}).nonstrict();


const $3956dd9640f2b520$export$c1a4367d4847eb06 = (rootDir)=>{
    const configPath = `${rootDir}/evidence.plugins.yaml`;
    try {
        const configFileContent = (0, ($parcel$interopDefault($1CGwI$fs))).readFileSync(configPath, "utf8").toString();
        // Surround all YAML key that begin with "@" in quotes
        // Skipping keys that are already quoted (e.g. beginning of line or whitespace)
        const rawConfig = (0, ($parcel$interopDefault($1CGwI$yaml))).parse(configFileContent.replaceAll(/($|\s)(@.+):/g, '$1"$2":'));
        const configResult = (0, $7b0083e9d7491326$export$e691085fbd9bf5be).safeParse(rawConfig);
        if (!configResult.success) {
            console.error((0, ($parcel$interopDefault($1CGwI$chalk))).bold.red(`[!] evidence.plugins.yaml does not contain a valid configuration. \n    Plugins will not be loaded. This may lead to unexpected behavior.`));
            const formattedError = (0, $63f219b28f6851a4$export$71e7d3deffa0730b)(configResult.error.format());
            console.error((0, ($parcel$interopDefault($1CGwI$chalk))).red("|   Discovered Errors:"));
            const redPipe = (0, ($parcel$interopDefault($1CGwI$chalk))).red("|");
            console.error(`${redPipe}   ${(0, ($parcel$interopDefault($1CGwI$yaml))).stringify(formattedError).replace(/\n/g, `\n${redPipe}   `)}`);
            throw new Error("Invalid evidence.plugins.yaml");
        }
        return configResult.data;
    } catch (e) {
        if (!(e instanceof Error)) throw e;
        if (e.message.startsWith("ENOENT")) throw new Error(`Could not find evidence plugins file. (Look at ${configPath})`, {
            cause: e
        });
        throw e;
    }
};
