// @evidence-dev/tailwind/dist/index.mjs
import {fontFamily as $lHQC2$fontFamily} from "tailwindcss/defaultTheme";

const $7e0ea4c5113f7428$export$8f45430ccf837300 = {
    color: {
        1: "#923d59",
        2: "#488f96",
        3: "#518eca",
        4: "#B3A9A0",
        5: "#3F6C51",
        6: "#ffc857",
        7: "#495867",
        8: "#bfdbf7",
        9: "#bc4749",
        10: "#EEEBD0",
        11: "#F68E5F",
        12: "#923d59",
        13: "#488f96",
        14: "#518eca",
        15: "#B3A9A0",
        16: "#3F6C51",
        17: "#ffc857",
        18: "#495867",
        19: "#bfdbf7",
        20: "#bc4749"
    }
};

const $7e0ea4c5113f7428$export$e506a1d27d1eaa20 = {
    theme: {
        extend: {
            colors: $7e0ea4c5113f7428$export$8f45430ccf837300,
            fontFamily: {
                sans: [
                    `Inter, ${$lHQC2$fontFamily.sans}`,
                    {
                        // open fours by default
                        fontFeatureSettings: '"cv02"'
                    }
                ],
                serif: [
                    `Spectral, ${$lHQC2$fontFamily.serif}`,
                    {
                        // proportional oldstyle figures by default
                        fontFeatureSettings: '"pnum", "onum"'
                    }
                ],
                mono: [
                    `'Menlo', 'Monaco', ${$lHQC2$fontFamily.mono}`
                ]
            },
            textShadow: {
                DEFAULT: "1px solid var(--tw-shadow-color)"
            },
            gridTemplateRows: {
                auto: "auto"
            },
            listStyleType: {
                circle: "circle",
                square: "square",
                "lower-alpha": "lower-alpha",
                "lower-roman": "lower-roman"
            }
        },
        plugins: []
    }
};
var $7e0ea4c5113f7428$export$2e2bcd8739ae039 = $7e0ea4c5113f7428$export$e506a1d27d1eaa20;


export {$7e0ea4c5113f7428$export$8f45430ccf837300 as colors, $7e0ea4c5113f7428$export$e506a1d27d1eaa20 as config, $7e0ea4c5113f7428$export$2e2bcd8739ae039 as default};
