// bits-ui/dist/bits/date-picker/ctx.js
import { createDatePicker } from "@melt-ui/svelte";
import { getContext, setContext } from "svelte";
import { removeUndefined, getOptionUpdater } from "../../internal/index.js";
import { getAttrs as getCalendarAttrs } from "../calendar/ctx.js";
import { getAttrs as getFieldAttrs } from "../date-field/ctx.js";
import { getAttrs as getPopoverAttrs } from "../popover/ctx.js";
import { getPositioningUpdater } from "../floating/helpers.js";
const NAME = "date-picker";
export { getCalendarAttrs, getFieldAttrs, getPopoverAttrs };
export function setCtx(props) {
    const datePicker = createDatePicker(removeUndefined(props));
    const updateOption = getOptionUpdater(datePicker.options);
    setContext(NAME, { ...datePicker, updateOption });
    return {
        ...datePicker,
        updateOption: getOptionUpdater(datePicker.options)
    };
}
export function getCtx() {
    return getContext(NAME);
}
const defaultPlacement = {
    side: "bottom",
    align: "center"
};
export function updatePositioning(props) {
    const withDefaults = { ...defaultPlacement, ...props };
    const { options: { positioning } } = getCtx();
    const updater = getPositioningUpdater(positioning);
    updater(withDefaults);
}
